File: ClipmanIOS/ClipmanIOS/Core/SyncConflictResolver.swift
import CryptoKit
import Foundation

struct OptimisticEntryDeletion: Sendable {
    let previousDatabase: ClipDatabase
    let optimisticDatabase: ClipDatabase

    init?(database: ClipDatabase, entryID: String, machineName: String) {
        guard let entryIndex = database.Entries.firstIndex(where: { $0.Id == entryID }) else { return nil }
        previousDatabase = database
        var updated = database
        let entry = updated.Entries.remove(at: entryIndex)
        let now = TimeUtil.nowUnixMs()
        let marker = DeletedClipEntry(
            Id: entry.Id,
            TextHash: SyncConflictResolver.textHash(entry.Text),
            DeletedUnixMs: now,
            SourceMachine: machineName
        )
        if let markerIndex = updated.DeletedEntries.firstIndex(where: { $0.Id == entry.Id }) {
            updated.DeletedEntries[markerIndex] = marker
        } else {
            updated.DeletedEntries.append(marker)
        }
        updated.Version = max(1, updated.Version)
        updated.UpdatedUnixMs = now
        for index in updated.Entries.indices {
            updated.Entries[index].ManualOrder = Int64(index + 1)
        }
        optimisticDatabase = updated
    }

    func restoredDatabase(ifCurrentMatches current: ClipDatabase) -> ClipDatabase? {
        guard SyncConflictResolver.hasSameContent(current, optimisticDatabase) else { return nil }
        return previousDatabase
    }
}

enum SyncConflictResolver {
    static func hasSameContent(_ left: ClipDatabase, _ right: ClipDatabase) -> Bool {
        left.Entries == right.Entries && left.DeletedEntries == right.DeletedEntries
    }

    static func merge(target: ClipDatabase, source: ClipDatabase) -> ClipDatabase {
        var merged = target
        mergeDeletedEntries(into: &merged, source: source)
        applyDeletedEntries(&merged)
        let deletionIndex = DeletionIndex(merged.DeletedEntries)
        var entryIndexesByID: [String: Int] = [:]
        var entryIndexesByText: [String: Int] = [:]
        for index in merged.Entries.indices {
            let entry = merged.Entries[index]
            if !entry.Id.isEmpty, entryIndexesByID[entry.Id.lowercased()] == nil {
                entryIndexesByID[entry.Id.lowercased()] = index
            }
            if entryIndexesByText[entry.Text] == nil {
                entryIndexesByText[entry.Text] = index
            }
        }

        for incoming in source.Entries where !incoming.Text.isEmpty {
            if deletionIndex.contains(incoming) { continue }
            if !incoming.Id.isEmpty, let idIndex = entryIndexesByID[incoming.Id.lowercased()] {
                let previousText = merged.Entries[idIndex].Text
                mergeEntry(existing: &merged.Entries[idIndex], incoming: incoming)
                if previousText != merged.Entries[idIndex].Text {
                    entryIndexesByText.removeValue(forKey: previousText)
                }
                if entryIndexesByText[merged.Entries[idIndex].Text] == nil {
                    entryIndexesByText[merged.Entries[idIndex].Text] = idIndex
                }
                continue
            }
            if let textIndex = entryIndexesByText[incoming.Text] {
                mergeEntry(existing: &merged.Entries[textIndex], incoming: incoming)
                if !merged.Entries[textIndex].Id.isEmpty,
                   entryIndexesByID[merged.Entries[textIndex].Id.lowercased()] == nil {
                    entryIndexesByID[merged.Entries[textIndex].Id.lowercased()] = textIndex
                }
                entryIndexesByText[merged.Entries[textIndex].Text] = textIndex
                continue
            }
            merged.Entries.append(incoming)
            let index = merged.Entries.count - 1
            if !incoming.Id.isEmpty {
                entryIndexesByID[incoming.Id.lowercased()] = index
            }
            entryIndexesByText[incoming.Text] = index
        }
        applyDeletedEntries(&merged)
        normalize(&merged)
        return merged
    }

    static func addText(database: ClipDatabase, text: String, machineName: String, richText: RichTextPayload? = nil) -> ClipDatabase {
        let trimmed = text.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return database }
        var result = database
        let now = TimeUtil.nowUnixMs()
        let normalizedRichText = MobileRichTextClipboard.normalize(richText)
        if let index = result.Entries.firstIndex(where: { $0.Text == trimmed }) {
            result.Entries[index].LastUsedUnixMs = now
            result.Entries[index].SourceMachine = machineName
            if let normalizedRichText {
                result.Entries[index].RichText = normalizedRichText
                result.Entries[index].RichTextUpdatedUnixMs = now
            }
        } else {
            let nextOrder = (result.Entries.map(\.ManualOrder).max() ?? 0) + 1
            result.Entries.append(ClipEntry(
                Text: trimmed,
                SourceMachine: machineName,
                CreatedUnixMs: now,
                LastUsedUnixMs: now,
                ModifiedUnixMs: now,
                ManualOrder: nextOrder,
                RichText: normalizedRichText,
                RichTextUpdatedUnixMs: normalizedRichText == nil ? 0 : now
            ))
        }
        normalize(&result)
        return result
    }

    static func updateEntry(database: ClipDatabase, entry: ClipEntry, machineName: String) -> ClipDatabase {
        var result = database
        guard let index = result.Entries.firstIndex(where: { $0.Id == entry.Id }) else { return result }
        var updated = entry
        updated.Text = updated.Text.trimmingCharacters(in: .whitespacesAndNewlines)
        updated.Name = updated.Name.trimmingCharacters(in: .whitespacesAndNewlines)
        updated.Group = updated.Group.trimmingCharacters(in: .whitespacesAndNewlines)
        updated.SourceMachine = machineName
        updated.LastUsedUnixMs = TimeUtil.nowUnixMs()
        updated.ModifiedUnixMs = updated.LastUsedUnixMs
        if updated.Text != result.Entries[index].Text {
            updated.RichText = nil
            updated.RichTextUpdatedUnixMs = updated.LastUsedUnixMs
        }
        if updated.Text.isEmpty {
            return deleteEntry(database: database, entryID: entry.Id, machineName: machineName)
        }
        result.Entries[index] = updated
        normalize(&result)
        return result
    }

    static func togglePinned(database: ClipDatabase, entryID: String) -> ClipDatabase {
        var result = database
        guard let index = result.Entries.firstIndex(where: { $0.Id == entryID }) else { return result }
        result.Entries[index].Pinned.toggle()
        result.Entries[index].LastUsedUnixMs = TimeUtil.nowUnixMs()
        result.Entries[index].ModifiedUnixMs = result.Entries[index].LastUsedUnixMs
        normalize(&result)
        return result
    }

    static func deleteEntry(database: ClipDatabase, entryID: String, machineName: String) -> ClipDatabase {
        var result = database
        guard let entry = result.Entries.first(where: { $0.Id == entryID }) else { return result }
        let marker = DeletedClipEntry(Id: entry.Id, TextHash: textHash(entry.Text), DeletedUnixMs: TimeUtil.nowUnixMs(), SourceMachine: machineName)
        result.Entries.removeAll { $0.Id == entryID }
        if let index = result.DeletedEntries.firstIndex(where: { $0.Id == marker.Id }) {
            result.DeletedEntries[index] = marker
        } else {
            result.DeletedEntries.append(marker)
        }
        normalize(&result)
        return result
    }

    static func textHash(_ text: String) -> String {
        let digest = SHA256.hash(data: Data(text.utf8))
        return digest.map { String(format: "%02x", $0) }.joined()
    }

    private static func mergeDeletedEntries(into target: inout ClipDatabase, source: ClipDatabase) {
        normalizeDeletedEntries(&target)
        var normalizedSource = source
        normalizeDeletedEntries(&normalizedSource)
        var byID = Dictionary(uniqueKeysWithValues: target.DeletedEntries.map { ($0.Id, $0) })
        for deleted in normalizedSource.DeletedEntries where !deleted.Id.isEmpty {
            if var existing = byID[deleted.Id] {
                if deleted.DeletedUnixMs > existing.DeletedUnixMs {
                    byID[deleted.Id] = deleted
                } else if existing.TextHash.isEmpty && !deleted.TextHash.isEmpty {
                    existing.TextHash = deleted.TextHash
                    byID[deleted.Id] = existing
                }
            } else {
                byID[deleted.Id] = deleted
            }
        }
        target.DeletedEntries = Array(byID.values)
        normalizeDeletedEntries(&target)
    }

    private static func applyDeletedEntries(_ database: inout ClipDatabase) {
        guard !database.DeletedEntries.isEmpty else { return }
        let deletionIndex = DeletionIndex(database.DeletedEntries)
        database.Entries.removeAll { deletionIndex.contains($0) }
    }

    private static func normalize(_ database: inout ClipDatabase) {
        database.Version = max(1, database.Version)
        database.UpdatedUnixMs = TimeUtil.nowUnixMs()
        normalizeDeletedEntries(&database)
        applyDeletedEntries(&database)
        database.Entries = database.Entries
            .filter { !$0.Text.isEmpty }
            .sorted {
                let leftOrder = $0.ManualOrder <= 0 ? Int64.max : $0.ManualOrder
                let rightOrder = $1.ManualOrder <= 0 ? Int64.max : $1.ManualOrder
                if leftOrder == rightOrder { return $0.CreatedUnixMs < $1.CreatedUnixMs }
                return leftOrder < rightOrder
            }
        for index in database.Entries.indices {
            if database.Entries[index].Id.isEmpty {
                database.Entries[index].Id = UUID().uuidString.replacingOccurrences(of: "-", with: "").lowercased()
            }
            database.Entries[index].ManualOrder = Int64(index + 1)
        }
    }

    private static func normalizeDeletedEntries(_ database: inout ClipDatabase) {
        let cutoff = TimeUtil.nowUnixMs() - Int64(90 * 24 * 60 * 60 * 1_000)
        var byID: [String: DeletedClipEntry] = [:]
        for marker in database.DeletedEntries {
            let id = marker.Id.trimmingCharacters(in: .whitespacesAndNewlines)
            guard !id.isEmpty else { continue }
            guard marker.DeletedUnixMs == 0 || marker.DeletedUnixMs >= cutoff else { continue }
            var normalized = marker
            normalized.Id = id
            normalized.TextHash = normalized.TextHash.trimmingCharacters(in: .whitespacesAndNewlines)
            if normalized.DeletedUnixMs == 0 {
                normalized.DeletedUnixMs = TimeUtil.nowUnixMs()
            }
            if let existing = byID[id], existing.DeletedUnixMs >= normalized.DeletedUnixMs {
                if existing.TextHash.isEmpty && !normalized.TextHash.isEmpty {
                    var repaired = existing
                    repaired.TextHash = normalized.TextHash
                    byID[id] = repaired
                }
                continue
            }
            byID[id] = normalized
        }
        database.DeletedEntries = byID.values.sorted {
            if $0.DeletedUnixMs == $1.DeletedUnixMs { return $0.Id < $1.Id }
            return $0.DeletedUnixMs > $1.DeletedUnixMs
        }
    }

    private struct DeletionIndex {
        private let deletedIDs: Set<String>
        private let latestDeletionByTextHash: [String: Int64]

        init(_ deletedEntries: [DeletedClipEntry]) {
            deletedIDs = Set(deletedEntries.lazy.map(\.Id).filter { !$0.isEmpty })
            var byHash: [String: Int64] = [:]
            for marker in deletedEntries where !marker.TextHash.isEmpty {
                byHash[marker.TextHash] = max(byHash[marker.TextHash] ?? Int64.min, marker.DeletedUnixMs)
            }
            latestDeletionByTextHash = byHash
        }

        func contains(_ entry: ClipEntry) -> Bool {
            if deletedIDs.contains(entry.Id) { return true }
            guard !entry.Text.isEmpty else { return false }
            let hash = SyncConflictResolver.textHash(entry.Text)
            guard let deletedUnixMs = latestDeletionByTextHash[hash] else { return false }
            let entryChangedUnixMs = max(entry.CreatedUnixMs, entry.LastUsedUnixMs)
            return deletedUnixMs <= 0 || entryChangedUnixMs <= deletedUnixMs
        }
    }

    private static func mergeEntry(existing: inout ClipEntry, incoming: ClipEntry) {
        let incomingWins = incoming.LastUsedUnixMs >= existing.LastUsedUnixMs
        let incomingCreatedWins = incoming.CreatedUnixMs > existing.CreatedUnixMs
        let incomingModifiedWins = incoming.ModifiedUnixMs > existing.ModifiedUnixMs
        let bothLegacy = incoming.ModifiedUnixMs <= 0 && existing.ModifiedUnixMs <= 0
        let legacyTextRepair = bothLegacy
            && !existing.Id.isEmpty
            && existing.Id.caseInsensitiveCompare(incoming.Id) == .orderedSame
            && existing.Text != incoming.Text
        if incoming.LastUsedUnixMs > existing.LastUsedUnixMs {
            existing.LastUsedUnixMs = incoming.LastUsedUnixMs
        }
        if incoming.CreatedUnixMs > 0 && (existing.CreatedUnixMs == 0 || incomingCreatedWins || (!incomingWins && incoming.CreatedUnixMs < existing.CreatedUnixMs)) {
            existing.CreatedUnixMs = incoming.CreatedUnixMs
        }
        if incomingModifiedWins {
            let textChanged = existing.Text != incoming.Text
            existing.Text = incoming.Text
            existing.Name = incoming.Name.trimmingCharacters(in: .whitespacesAndNewlines)
            existing.Group = incoming.Group.trimmingCharacters(in: .whitespacesAndNewlines)
            existing.Pinned = incoming.Pinned
            existing.IsTemplate = incoming.IsTemplate
            existing.ManualOrder = incoming.ManualOrder
            existing.ModifiedUnixMs = incoming.ModifiedUnixMs
            if textChanged {
                existing.RichText = incoming.RichText
                existing.RichTextUpdatedUnixMs = max(incoming.RichTextUpdatedUnixMs, incoming.ModifiedUnixMs)
            }
        } else if bothLegacy {
            if legacyTextRepair {
                existing.Text = incoming.Text
                existing.RichText = incoming.RichText
                existing.RichTextUpdatedUnixMs = max(existing.RichTextUpdatedUnixMs, incoming.RichTextUpdatedUnixMs)
            }
            if !incoming.Name.isEmpty && incomingWins {
                existing.Name = incoming.Name
            }
            if !incoming.Group.isEmpty && incomingWins {
                existing.Group = incoming.Group
            }
            existing.Pinned = existing.Pinned || incoming.Pinned
            existing.IsTemplate = existing.IsTemplate || incoming.IsTemplate
            if existing.ManualOrder <= 0 || (incoming.ManualOrder > 0 && incoming.ManualOrder < existing.ManualOrder) {
                existing.ManualOrder = incoming.ManualOrder
            }
        }
        if !incoming.SourceMachine.isEmpty && (incomingWins || incomingCreatedWins) {
            existing.SourceMachine = incoming.SourceMachine
        }
        if incoming.RichTextUpdatedUnixMs > existing.RichTextUpdatedUnixMs {
            existing.RichText = incoming.RichText
            existing.RichTextUpdatedUnixMs = incoming.RichTextUpdatedUnixMs
        }
    }
}

File: ClipmanAndroid/app/src/main/java/me/onj/clipman/SyncConflictResolver.kt
package me.onj.clipman

import java.security.MessageDigest
import java.util.UUID

object SyncConflictResolver {
    fun merge(target: ClipDatabase, source: ClipDatabase): ClipDatabase {
        val deletedById = normalizeDeletedEntries(target.DeletedEntries + source.DeletedEntries)
            .filter { it.Id.isNotBlank() }
            .groupBy { it.Id }
            .mapValues { (_, markers) -> markers.maxBy { it.DeletedUnixMs } }
            .values
            .toList()
        val deletionIndex = DeletionIndex(deletedById)
        val retainedTarget = target.Entries.filterNot { deletionIndex.contains(it) }
        val byId = retainedTarget.associateBy { it.Id }.toMutableMap()
        val byText = retainedTarget.associateBy { it.Text }.toMutableMap()
        val merged = retainedTarget.toMutableList()

        for (incoming in source.Entries) {
            if (incoming.Text.isEmpty()) continue
            if (deletionIndex.contains(incoming)) continue
            val existing = byId[incoming.Id].takeUnless { incoming.Id.isBlank() } ?: byText[incoming.Text]
            if (existing == null) {
                val normalized = incoming.normalized()
                merged.add(normalized)
                if (normalized.Id.isNotBlank()) byId[normalized.Id] = normalized
                byText[normalized.Text] = normalized
            } else {
                val index = merged.indexOfFirst { it === existing || it.Id == existing.Id || it.Text == existing.Text }
                val previousText = existing.Text
                val updated = mergeEntry(existing, incoming)
                if (index >= 0) merged[index] = updated
                if (updated.Id.isNotBlank()) byId[updated.Id] = updated
                if (previousText != updated.Text) byText.remove(previousText)
                byText[updated.Text] = updated
            }
        }

        return normalize(target.copy(Entries = merged, DeletedEntries = deletedById))
    }

    fun hasSameContent(left: ClipDatabase, right: ClipDatabase): Boolean =
        left.Entries == right.Entries && left.DeletedEntries == right.DeletedEntries

    fun addText(database: ClipDatabase, text: String, machineName: String, richText: RichTextPayload? = null): ClipDatabase {
        val trimmed = text.trim()
        if (trimmed.isEmpty()) return database
        val now = TimeUtil.nowUnixMs()
        val existing = database.Entries.firstOrNull { it.Text == trimmed }
        val normalizedRichText = RichTextClipboard.normalize(richText)
        val storedRichText = normalizedRichText ?: existing?.RichText
        val richTextUpdated = if (normalizedRichText == null) existing?.RichTextUpdatedUnixMs ?: 0 else now
        val withoutDuplicate = database.Entries.filterNot { it.Text == trimmed }
        val nextManualOrder = withoutDuplicate.maxOfOrNull { it.ManualOrder }?.plus(1) ?: 1
        val entry = ClipEntry(
            Id = UUID.randomUUID().toString().replace("-", ""),
            Text = trimmed,
            SourceMachine = machineName,
            CreatedUnixMs = now,
            LastUsedUnixMs = now,
            ModifiedUnixMs = now,
            ManualOrder = nextManualOrder,
            RichText = storedRichText,
            RichTextUpdatedUnixMs = richTextUpdated
        )
        return normalize(database.copy(Entries = withoutDuplicate + entry, UpdatedUnixMs = now))
    }

    fun updateEntry(database: ClipDatabase, entry: ClipEntry): ClipDatabase {
        if (entry.Id.isBlank()) return database
        val now = TimeUtil.nowUnixMs()
        val requestedGroup = entry.Group.trim()
        val normalizedGroup = canonicalGroup(database.Entries, requestedGroup)
        val updated = entry.copy(
            Text = entry.Text.trim(),
            Name = entry.Name.trim(),
            Group = normalizedGroup,
            SourceMachine = entry.SourceMachine.trim(),
            LastUsedUnixMs = now,
            ModifiedUnixMs = now,
            RichText = if (entry.Text.trim() == database.Entries.firstOrNull { it.Id == entry.Id }?.Text) entry.RichText else null,
            RichTextUpdatedUnixMs = if (entry.Text.trim() == database.Entries.firstOrNull { it.Id == entry.Id }?.Text) entry.RichTextUpdatedUnixMs else now
        )
        if (updated.Text.isBlank()) return deleteEntry(database, entry.Id)
        return normalize(database.copy(
            Entries = database.Entries.map { if (it.Id == entry.Id) updated else it },
            UpdatedUnixMs = now
        ))
    }

    private fun canonicalGroup(entries: List<ClipEntry>, requested: String): String {
        if (requested.isBlank()) return ""
        return entries
            .filter { it.Group.trim().equals(requested, ignoreCase = true) }
            .groupBy { it.Group.trim() }
            .map { (label, spelling) ->
                Triple(
                    label,
                    spelling.size,
                    spelling.maxOf { maxOf(it.ModifiedUnixMs, it.LastUsedUnixMs, it.CreatedUnixMs) }
                )
            }
            .sortedWith(
                compareByDescending<Triple<String, Int, Long>> { it.second }
                    .thenByDescending { it.third }
                    .thenBy { it.first }
            )
            .firstOrNull()?.first ?: requested
    }

    fun togglePinned(database: ClipDatabase, entryId: String): ClipDatabase {
        val now = TimeUtil.nowUnixMs()
        return normalize(database.copy(
            Entries = database.Entries.map {
                if (it.Id == entryId) it.copy(Pinned = !it.Pinned, LastUsedUnixMs = now, ModifiedUnixMs = now) else it
            },
            UpdatedUnixMs = now
        ))
    }

    fun deleteEntry(database: ClipDatabase, entryId: String): ClipDatabase {
        val entry = database.Entries.firstOrNull { it.Id == entryId } ?: return database
        val now = TimeUtil.nowUnixMs()
        val deleted = DeletedClipEntry(
            Id = entry.Id,
            TextHash = textHash(entry.Text),
            DeletedUnixMs = now,
            SourceMachine = entry.SourceMachine
        )
        return normalize(database.copy(
            Entries = database.Entries.filterNot { it.Id == entryId },
            DeletedEntries = database.DeletedEntries + deleted,
            UpdatedUnixMs = now
        ))
    }

    fun textHash(text: String): String {
        val digest = MessageDigest.getInstance("SHA-256").digest((text.ifEmpty { "" }).toByteArray(Charsets.UTF_8))
        return digest.joinToString("") { "%02x".format(it) }
    }

    private class DeletionIndex(deletedEntries: List<DeletedClipEntry>) {
        private val deletedIds = deletedEntries.mapNotNullTo(mutableSetOf()) { marker ->
            marker.Id.takeIf { it.isNotBlank() }
        }
        private val latestDeletionByTextHash = buildMap {
            deletedEntries.forEach { marker ->
                if (marker.TextHash.isNotBlank()) {
                    put(marker.TextHash, maxOf(get(marker.TextHash) ?: Long.MIN_VALUE, marker.DeletedUnixMs))
                }
            }
        }

        fun contains(entry: ClipEntry): Boolean {
            if (entry.Id in deletedIds) return true
            if (entry.Text.isEmpty()) return false
            val deletedUnixMs = latestDeletionByTextHash[textHash(entry.Text)] ?: return false
            val entryChangedUnixMs = maxOf(entry.CreatedUnixMs, entry.LastUsedUnixMs)
            return deletedUnixMs <= 0 || entryChangedUnixMs <= deletedUnixMs
        }
    }

    private fun normalizeDeletedEntries(markers: List<DeletedClipEntry>): List<DeletedClipEntry> {
        val now = TimeUtil.nowUnixMs()
        val cutoff = now - 90L * 24 * 60 * 60 * 1_000
        return markers
            .asSequence()
            .map { marker ->
                marker.copy(
                    Id = marker.Id.trim(),
                    TextHash = marker.TextHash.trim(),
                    DeletedUnixMs = if (marker.DeletedUnixMs == 0L) now else marker.DeletedUnixMs
                )
            }
            .filter { it.Id.isNotEmpty() && it.DeletedUnixMs >= cutoff }
            .toList()
    }

    private fun mergeEntry(existing: ClipEntry, incoming: ClipEntry): ClipEntry {
        val incomingWins = incoming.LastUsedUnixMs >= existing.LastUsedUnixMs
        val incomingCreatedWins = incoming.CreatedUnixMs > existing.CreatedUnixMs
        val incomingModifiedWins = incoming.ModifiedUnixMs > existing.ModifiedUnixMs
        val bothLegacy = incoming.ModifiedUnixMs <= 0 && existing.ModifiedUnixMs <= 0
        val legacyTextRepair = bothLegacy && existing.Id.isNotBlank() && existing.Id.equals(incoming.Id, ignoreCase = true) && existing.Text != incoming.Text
        val textChangedByMerge = incomingModifiedWins && existing.Text != incoming.Text
        return existing.copy(
            Text = when {
                incomingModifiedWins || legacyTextRepair -> incoming.Text
                else -> existing.Text
            },
            CreatedUnixMs = when {
                incoming.CreatedUnixMs == 0L -> existing.CreatedUnixMs
                existing.CreatedUnixMs == 0L -> incoming.CreatedUnixMs
                incomingCreatedWins -> incoming.CreatedUnixMs
                !incomingWins && incoming.CreatedUnixMs < existing.CreatedUnixMs -> incoming.CreatedUnixMs
                else -> existing.CreatedUnixMs
            },
            LastUsedUnixMs = maxOf(existing.LastUsedUnixMs, incoming.LastUsedUnixMs),
            ModifiedUnixMs = if (incomingModifiedWins) incoming.ModifiedUnixMs else existing.ModifiedUnixMs,
            Name = when {
                incomingModifiedWins -> incoming.Name.trim()
                bothLegacy && incoming.Name.isNotBlank() && incomingWins -> incoming.Name.trim()
                else -> existing.Name
            },
            Group = when {
                incomingModifiedWins -> incoming.Group.trim()
                bothLegacy && incoming.Group.isNotBlank() && incomingWins -> incoming.Group.trim()
                else -> existing.Group
            },
            SourceMachine = if (incoming.SourceMachine.isNotBlank() && (incomingWins || incomingCreatedWins)) incoming.SourceMachine.trim() else existing.SourceMachine,
            Pinned = if (incomingModifiedWins) incoming.Pinned else if (bothLegacy) existing.Pinned || incoming.Pinned else existing.Pinned,
            IsTemplate = if (incomingModifiedWins) incoming.IsTemplate else if (bothLegacy) existing.IsTemplate || incoming.IsTemplate else existing.IsTemplate,
            RichText = when {
                textChangedByMerge -> incoming.RichText
                legacyTextRepair -> incoming.RichText
                incoming.RichTextUpdatedUnixMs > existing.RichTextUpdatedUnixMs -> incoming.RichText
                else -> existing.RichText
            },
            RichTextUpdatedUnixMs = when {
                textChangedByMerge -> maxOf(incoming.RichTextUpdatedUnixMs, incoming.ModifiedUnixMs)
                legacyTextRepair -> maxOf(existing.RichTextUpdatedUnixMs, incoming.RichTextUpdatedUnixMs)
                else -> maxOf(existing.RichTextUpdatedUnixMs, incoming.RichTextUpdatedUnixMs)
            },
            ManualOrder = when {
                incomingModifiedWins -> incoming.ManualOrder
                bothLegacy && existing.ManualOrder <= 0L -> incoming.ManualOrder
                bothLegacy && incoming.ManualOrder > 0L && incoming.ManualOrder < existing.ManualOrder -> incoming.ManualOrder
                else -> existing.ManualOrder
            }
        )
    }

    private fun normalize(database: ClipDatabase): ClipDatabase {
        var order = 1L
        val normalized = database.Entries
            .filter { it.Text.isNotEmpty() }
            .sortedWith(compareBy<ClipEntry> { if (it.ManualOrder <= 0) Long.MAX_VALUE else it.ManualOrder }.thenBy { it.CreatedUnixMs })
            .map { it.normalized(order++) }
        return database.copy(
            Version = maxOf(1, database.Version),
            UpdatedUnixMs = TimeUtil.nowUnixMs(),
            Entries = normalized,
            DeletedEntries = normalizeDeletedEntries(database.DeletedEntries)
        )
    }

    private fun ClipEntry.normalized(manualOrder: Long = ManualOrder): ClipEntry {
        val now = TimeUtil.nowUnixMs()
        return copy(
            Id = Id.ifBlank { UUID.randomUUID().toString().replace("-", "") },
            Text = Text,
            Name = Name,
            Group = Group,
            SourceMachine = SourceMachine,
            CreatedUnixMs = if (CreatedUnixMs == 0L) now else CreatedUnixMs,
            LastUsedUnixMs = if (LastUsedUnixMs == 0L) (if (CreatedUnixMs == 0L) now else CreatedUnixMs) else LastUsedUnixMs,
            ManualOrder = manualOrder
        )
    }
}

class ServerConflictException(message: String) : Exception(message)
